domain/agentprovisioner/service/service_test.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/errors"
+	"github.com/juju/proxy"
+	jc "github.com/juju/testing/checkers"
+	"go.uber.org/mock/gomock"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/core/containermanager"
+	"github.com/juju/juju/core/instance"
+	"github.com/juju/juju/core/model"
+	modeltesting "github.com/juju/juju/core/model/testing"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/rpc/params"
+)
+
+type suite struct {
+	state          *MockState
+	provider       *MockProvider
+	providerGetter func(context.Context) (Provider, error)
+	modelID        model.UUID
+	service        *Service
+}
+
+var _ = gc.Suite(&suite{})
+
+func (s *suite) SetUpTest(c *gc.C) {
+	s.modelID = modeltesting.GenModelUUID(c)
+	// Default provider getter function
+	s.providerGetter = func(ctx context.Context) (Provider, error) {
+		return s.provider, nil
+	}
+}
+
+func (s *suite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+	s.state = NewMockState(ctrl)
+	s.provider = NewMockProvider(ctrl)
+	return ctrl
+}
+
+func (s *suite) setupService(c *gc.C) {
+	s.service = NewService(s.modelID, s.state, s.providerGetter)
+}
+
+// TestContainerManagerConfigForType asserts the happy path for
+// Service.ContainerManagerConfigForType.
+func (s *suite) TestContainerManagerConfigForType(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	s.setupService(c)
+
+	s.state.EXPECT().GetModelConfigKeyValues(gomock.Any(), gomock.Any()).Return(map[string]string{
+		config.LXDSnapChannel:                            "5.0/stable",
+		config.ContainerImageMetadataURLKey:              "https://images.linuxcontainers.org/",
+		config.ContainerImageMetadataDefaultsDisabledKey: "true",
+		config.ContainerImageStreamKey:                   "released",
+	}, nil)
+	s.provider.EXPECT().SupportsContainerAddresses(gomock.Any()).Return(false, nil)
+
+	cfg, err := s.service.ContainerManagerConfigForType(context.Background(), instance.LXD)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Check(cfg, jc.DeepEquals, containermanager.Config{
+		ImageMetadataURL:         "https://images.linuxcontainers.org/",
+		ImageStream:              "released",
+		LXDSnapChannel:           "5.0/stable",
+		MetadataDefaultsDisabled: true,
+		ModelID:                  s.modelID,
+		NetworkingMethod:         containermanager.NetworkingMethodLocal,
+	})
+}
+
+// TestDetermineNetworkingMethodUserDefined tests that if the user specifies a
+// container networking method in model config, this will be used in the
+// container manager config.
+func (s *suite) TestDetermineNetworkingMethodUserDefined(c *gc.C) {
```

Comment:
I don't think the way in which this test goes about getting the job done is useful. `determinContainerNetworkingMethod` is an implementation detail of how we calculate the value. What we care about is pre and post conditions. We want to see that when value x is set in model config the resultant value y equals foo.

My suggestion would be to stop testing the internal funcs of this package and test the package as a whole being the pre and post conditions.

Reply:
let's talk about this. My thinking is we specifically want to test how the container networking method is determined, rather than polluting this test with the whole container manager config stuff, which is sufficiently tested elsewhere.

Reply:
moot point now, as this internal func is now a first-class citizen